### PR TITLE
Bandit Party Component Sync

### LINKED ISSUE #794: Sync BanditPartyComponent
## Description

1. Sync the values in the **Synced** list below.
2. Create an integration test for all values (1 for server changing the value and 1 for client changing the value if applicable)
3. Create a command to test changing the value in game [see similar command](https://github.com/Bannerlord-Coop-Team/BannerlordCoop/blob/development/source/GameInterface/Services/Towns/Commands/TownDebugCommand.cs#L455)

Also see additional information at the bottom for more information and templates

**Key**
Server side only (only allow the running of the function on the server side)
Client side only (only allow the running of the function on the client side)
Client side sync (request the server to change the value, server then changes the value and replies to the client allowing the change)
Server side sync (only server allows running of the function and send to all clients the value changed and clients update the value/call original function)

**Synced**

| Method/Field Name | Sync Type |
|:-----------------:|:---------:|
| Hideout | Server side sync |
| IsBossParty | Server side sync |
| _relatedSettlement | Server side sync |

**Deferred**

None

**Externally Deferred**

None

**Non-Synced**
None

## Intended Design

**Message networking**
image
**Integration Testing**
image
## Location

Create a branch based from [development](https://github.com/Bannerlord-Coop-Team/BannerlordCoop)

## Related Issues

N/A

## Requirements

N/A

## Additional information

Using the [GameInterface Service Tempale](https://github.com/Bannerlord-Coop-Team/BannerlordCoop/tree/development/source/GameInterface/Services/Template) create commands to test the synced values.

For network communication use [Server Service Template](https://github.com/Bannerlord-Coop-Team/BannerlordCoop/tree/development/source/Coop.Core/Server/Services/Template) and [Client Service Template](https://github.com/Bannerlord-Coop-Team/BannerlordCoop/tree/development/source/Coop.Core/Client/Services/Template)

It is recommended to start with an integration test, you can create one using the [test template](https://github.com/Bannerlord-Coop-Team/BannerlordCoop/tree/development/source/Coop.IntegrationTests/Template).

## Definition of Done
- [ ] Class level comments exist for all new classes.
- [ ] XUnit tests exist for every method that does not require the game to be ran.

### PR DESCRIPTION
**Pull request by EgardA**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] All new classes have class-level documentation comments, if there are any at all
- [ ] Tests for the changes have been added (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation update
- [ ] Other... Please describe:

## What is the current behavior?

No bandit party component sync

## What is the new behavior?

Resolves #794 

bandit party component sync

### CODE REVIEW

**Comment by EgardA** [comment]
I left out _relatedSettlement on purpose, it is only assigned on creation and is never changed.

File: source/E2E.Tests/Services/MobileParties/BanditPartyComponentTests.cs [hunk 1]
```diff
@@ -72,4 +74,60 @@ public void ClientCreateParty_DoesNothing()
...
+    [Fact]
+    public void ClientUpdateParty_DoesNothing()
+    {
+        // Arrange
+        var server = TestEnvironment.Server;
+        var client1 = TestEnvironment.Clients.First();
+
+        var component = GameObjectCreator.CreateInitializedObject<BanditPartyComponent>();
+        var hideout = GameObjectCreator.CreateInitializedObject<Hideout>();
+        var settlement = GameObjectCreator.CreateInitializedObject<Settlement>();
```
Comment (garrettluskey) [inline source/E2E.Tests/Services/MobileParties/BanditPartyComponentTests.cs]: This isn't guaranteed to run on the server, could you add this in a server.call statement?

File: source/E2E.Tests/Util/ObjectBuilders/BanditPartyComponentBuilder.cs [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+namespace E2E.Tests.Util.ObjectBuilders;
+internal class BanditPartyComponentBuilder : IObjectBuilder
+{
+    public object Build()
+    {
+        var partyComponent = ObjectHelper.SkipConstructor<BanditPartyComponent>();
+        var settlement = GameObjectCreator.CreateInitializedObject<Settlement>();
+
+        AccessTools.Constructor(typeof(BanditPartyComponent), new Type[] { typeof(Settlement) })
+            .Invoke(partyComponent, new object[] { settlement });
```
Comment (garrettluskey) [inline source/E2E.Tests/Util/ObjectBuilders/BanditPartyComponentBuilder.cs]: Doesn't this work?

```suggestion
        return new BanditPartyComponent(settlement)
```

**Review by garrettluskey — CHANGES_REQUESTED** [review]

File: source/E2E.Tests/Services/MobileParties/BanditPartyComponentTests.cs [hunk 3]
```diff
@@ -72,4 +74,69 @@ public void ClientCreateParty_DoesNothing()
...
+        BanditPartyComponent component = GameObjectCreator.CreateInitializedObject<BanditPartyComponent>();
+        Hideout hideout = GameObjectCreator.CreateInitializedObject<Hideout>();
+        Settlement settlement = GameObjectCreator.CreateInitializedObject<Settlement>();
+
+        server.Call(() =>
+        {
+            component = GameObjectCreator.CreateInitializedObject<BanditPartyComponent>();
+            hideout = GameObjectCreator.CreateInitializedObject<Hideout>();
+            settlement = GameObjectCreator.CreateInitializedObject<Settlement>();
+        });
```
Comment (garrettluskey) [inline source/E2E.Tests/Services/MobileParties/BanditPartyComponentTests.cs]:
```suggestion
        string componentId = TestEnvironment.CreateRegisteredObject<BanditPartyComponent>();
        string hideoutId = TestEnvironment.CreateRegisteredObject<Hideout>();
        string settlementId = TestEnvironment.CreateRegisteredObject<Settlement>();
```

**Then use the ids to resolve the objects in the clients scope**

**Review by garrettluskey — CHANGES_REQUESTED** [review]

File: source/E2E.Tests/Services/PartyComponents/BanditPartyComponentTests.cs
Comment (georgyrudnev) [inline source/E2E.Tests/Services/PartyComponents/BanditPartyComponentTests.cs:1]: In lines 125-129, arent you just testing against the variables you created before?
Shouldnt you be testing against the client values through client.ObjectManager.TryGetObject ?

Comment (EgardA) [reply source/E2E.Tests/Services/PartyComponents/BanditPartyComponentTests.cs:1]: You are correct, I will fix that :)

**Pull request merged**